src/main/java/net/minecraftforge/event/entity/SpawnPlacementRegisterEvent.java

```diff
@@ -0,0 +1,207 @@
+     */
+    public void register(EntityType<?> entityType, @Nullable SpawnPlacements.Type placementType, @Nullable Heightmap.Types heightmap, SpawnPlacements.SpawnPredicate<Entity> predicate, Operation operation)
+    {
+        if (!map.containsKey(entityType))
+        {
+            if (placementType == null)
+            {
+                throw new NullPointerException("Registering a new Spawn Predicate requires a nonnull placement type! Entity Type: " + entityType.getClass().getSimpleName());
+            }
+            if (heightmap == null)
+            {
+                throw new NullPointerException("Registering a new Spawn Predicate requires a nonnull heightmap type! Entity Type: " + entityType.getClass().getSimpleName());
+            }
+            map.put(entityType, new MergedSpawnPredicate(predicate, placementType, heightmap));
+        }
+        else
+        {
+            map.get(entityType).merge(operation, predicate, placementType, heightmap);
+        }
+    }
+
+    public enum Operation
+    {
+        /**
+         * Checked third, these predicates must all pass along with the original predicate
+         */
+        AND,
+        /**
+         * Checked second, only one of these predicates must pass along with the original predicate
+         */
+        OR,
+        /**
+         * Checked first, the last mod to replace the predicate wipes out all other predicates. Listen with a low {@link EventPriority} if you need to do this.
+         */
+        REPLACE
+    }
+
+    @SuppressWarnings("unchecked")
+    public static class MergedSpawnPredicate
+    {
+        private final SpawnPlacements.SpawnPredicate<Entity> originalPredicate;
+        private final List<SpawnPlacements.SpawnPredicate<Entity>> orPredicates;
+        private final List<SpawnPlacements.SpawnPredicate<Entity>> andPredicates;
+        @Nullable private SpawnPlacements.SpawnPredicate<?> replacementPredicate;
+        private SpawnPlacements.Type spawnType;
+        private Heightmap.Types heightmapType;
+
+        public MergedSpawnPredicate(SpawnPlacements.SpawnPredicate<?> originalPredicate, SpawnPlacements.Type spawnType, Heightmap.Types heightmapType)
+        {
+            this.originalPredicate = (SpawnPlacements.SpawnPredicate<Entity>) originalPredicate;
+            this.orPredicates = new ArrayList<>();
+            this.andPredicates = new ArrayList<>();
+            this.replacementPredicate = null;
+            this.spawnType = spawnType;
+            this.heightmapType = heightmapType;
+        }
+
+        public SpawnPlacements.Type getSpawnType()
+        {
+            return spawnType;
+        }
+
+        public Heightmap.Types getHeightmapType()
+        {
+            return heightmapType;
+        }
+
+        private void merge(Operation operation, SpawnPlacements.SpawnPredicate<?> predicate, @Nullable SpawnPlacements.Type spawnType, @Nullable Heightmap.Types heightmapType)
```

Comment:
Should this method throw an error if the spawn and heightmap are nonnull when performing an AND or OR operation?